official/recommendation/data_pipeline.py

```diff
@@ -0,0 +1,727 @@
+
+    self._train_dataset.put(i, {
+        movielens.USER_COLUMN: users,
+        movielens.ITEM_COLUMN: items,
+        rconst.MASK_START_INDEX: np.array(mask_start_index, dtype=np.int32),
+        "labels": labels,
+    })
+
+  def _wait_to_construct_train_epoch(self):
+    count = 0
+    while self._train_dataset.buffer_reached() and not self._stop_loop:
+      time.sleep(0.01)
+      count += 1
+      if count >= 100 and np.log10(count) == np.round(np.log10(count)):
+        tf.logging.info(
+            "Waited {} times for training data to be consumed".format(count))
+
+  def _construct_training_epoch(self):
+    """Loop to construct a batch of training data."""
+    self._wait_to_construct_train_epoch()
+    start_time = timeit.default_timer()
+    if self._stop_loop:
+      return
+
+    self._train_dataset.start_construction()
+    map_args = list(range(self.train_batches_per_epoch))
+    assert not self._current_epoch_order.shape[0]
+    self._current_epoch_order = six.next(self._shuffle_iterator)
+
+    with popen_helper.get_threadpool(6) as pool:
+      pool.map(self._get_training_batch, map_args)
+    self._train_dataset.end_construction()
+
+    tf.logging.info("Epoch construction complete. Time: {:.1f} seconds".format(
+        timeit.default_timer() - start_time))
+
+  @staticmethod
+  def _assemble_eval_batch(users, positive_items, negative_items,
+                           users_per_batch):
+    """Construct duplicate_mask and structure data accordingly.
+
+    The positive items should be last so that they lose ties. However, they
+    should not be masked out if the true eval positive happens to be
+    selected as a negative. So instead, the positive is placed in the first
+    position, and then switched with the last element after the duplicate
+    mask has been computed.
+
+    Args:
+      users: An array of users in a batch. (should be identical along axis 1)
+      positive_items: An array (batch_size x 1) of positive item indices.
+      negative_items: An array of negative item indices.
+      users_per_batch: How many users should be in the batch. This is passed
+        as an argument so that ncf_test.py can use this method.
+
+    Returns:
+      User, item, and duplicate_mask arrays.
+    """
+    items = np.concatenate([positive_items, negative_items], axis=1)
+
+    # We pad the users and items here so that the duplicate mask calculation
+    # will include the padding. The metric function relies on every element
+    # except the positive being marked as duplicate to mask out padded points.
+    if users.shape[0] < users_per_batch:
+      pad_rows = users_per_batch - users.shape[0]
+      padding = np.zeros(shape=(pad_rows, users.shape[1]), dtype=np.int32)
+      users = np.concatenate([users, padding.astype(users.dtype)], axis=0)
+      items = np.concatenate([items, padding.astype(items.dtype)], axis=0)
+
+    duplicate_mask = stat_utils.mask_duplicates(items, axis=1).astype(np.bool)
```

Comment:
I think it does need to. You use the duplicated mask [here](https://github.com/tensorflow/models/blob/064c10e80543815a7425c890d32c83d082dfa8b8/official/recommendation/neumf_model.py#L379). If you do not mark it as a duplicate, it will be used in the HR and NDCG calculations.